open-craft/learner-profile-dashboard
# Basic UI, persist learner answers upon submit

cf. [OC-4159](https://tasks.opencraft.com/browse/OC-4159), [OC-4160](https://tasks.opencraft.com/browse/OC-4160)

This PR ships the following features:

- A basic UI for the LPD that is capable of rendering all types of questions that the MVP needs to support.
- Logic for storing answers to different types of questions in the DB.
- Logic for displaying saved answers back to the learner when revisiting the LPD.
- Functionality for expanding and collapsing individual sections of the learner profile.
- Logic that limits learner interactions with MRQs and Ranking questions according to the unique constraints of these question types.
    - MRQs don't allow learners to select more options than `MultipleChoiceQuestion.max_options_to_select`.
    - Rating questions don't allow learners to select the same rank for more than one option.

**Test instructions**

1. Run migrations.
1. Load test data for the `lpd` by running `./manage.py loaddata db-lpd.json` (see ticket attachments for the `db-lpd.json` file). If that doesn't work for some reason, launch the Django shell and `%cpaste` the commands from `commands.txt` (also attached to the ticket) to populate the DB.
1. Access the LPD in the LMS. (This assumes you've already [embedded](https://github.com/open-craft/learner-profile-dashboard/tree/27681e1a1379cee3b85b3135605129788e489d69#using-lpd-with-open-edx) the LPD as an LTI component in Open edX.)

1. Verify that you can collapse and expand individual sections of the LPD by clicking on their titles.

1. Refresh the page a few times. Observe that the order of options changes for questions that have `randomize_options` set to `True`. (When using test data from `db-lpd.json` or `commands.txt`, questions that are configured to display their options in random order include a multiple choice question ("Choose one of:") and a ranking question ("Rank these on a scale of 1 to 3:").) Observe that the order of options does **not** change for the remaining questions.

1. Answer a few questions and click "Submit your answers". Make sure to enter custom input for some of the answer options.

1. Refresh the page. Observe that the LPD displays the answers that you provided in the previous step, as well as any custom input you provided for specific answer options. This works for both regular refreshes and refreshes that [override the cache](https://support.mozilla.org/en-US/questions/1073264).

1. Verify that you can not select more than two options for the MRQ ("Choose at most two of:").
    - After selecting two options, checkboxes for the remaining (unselected) options will be disabled.
    - After deselecting one of the two selected options, all options will be enabled.
    - After refreshing the page with two options selected, checkboxes for the remaining (unselected) options will remain disabled.
    - After refreshing the page with a single option selected, all options will remain enabled.

1. Verify that you can not select the same rank for two different options belonging to the same ranking question. For example, if you select "1" for the "I love salad" option, selecting "1" for the "I love candy" option will result in "1" getting deselected for the "I love salad" option. The intended behavior for ranking questions is to never allow learners to rank more than `RatingQuestion.number_of_options_to_rank`, and for the value of each ranked option to be unique. (It is OK to rank *less* than `RatingQuestion.number_of_options_to_rank`, though.)

**Notes**

- This PR does not introduce custom CSS for the LPD. We will take care of that via a separate ticket.
- The build is currently failing due to a few leftover quality issues. To save time I will fix these after addressing review comments.
- The diff contains a few `print` and `console.log` statements that have been helpful for debugging purposes. I will remove these after addressing review comments.
- **Code review**: You can skip over third-party JS libraries added to `app/static/js/vendor/`.

**Reviewers**
